File: app/src/main/java/in/rahulja/getlogs/LogLocationWorker.java
package in.rahulja.getlogs;

import android.content.Context;
import android.content.Intent;
import android.location.Location;
import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.Tasks;
import java.util.concurrent.ExecutionException;
import androidx.core.app.ActivityCompat;
import android.Manifest;
import android.content.pm.PackageManager;

public class LogLocationWorker extends Worker {

    private static final String LOCATION_ACTION = "in.rahulja.getlogs.LAST_LOCATION";

    public LogLocationWorker(@NonNull Context context, @NonNull WorkerParameters params) {
        super(context, params);
    }

    @NonNull
    @Override
    public Result doWork() {
        Context context = getApplicationContext();
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
             broadcastLocation("ERROR_GETTING", "ERROR_GETTING");
             return Result.failure();
        }

        FusedLocationProviderClient fusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
        try {
            Location location = Tasks.await(fusedLocationClient.getLastLocation());
            if (location != null) {
                broadcastLocation(String.valueOf(location.getLatitude()), String.valueOf(location.getLongitude()));
            } else {
                 broadcastLocation(null, null);
            }
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            broadcastLocation("ERROR_GETTING", "ERROR_GETTING");
            return Result.failure();
        }

        return Result.success();
    }

    private void broadcastLocation(String lat, String lng) {
        Intent intent = new Intent();
        intent.setAction(LOCATION_ACTION);
        intent.putExtra("latitude", lat);
        intent.putExtra("longitude", lng);
        getApplicationContext().sendBroadcast(intent);
    }
}

File: app/src/main/java/in/rahulja/getlogs/service/LogLocationWorker.kt
package `in`.rahulja.getlogs.service

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import androidx.core.app.ActivityCompat
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import `in`.rahulja.getlogs.data.LogRepository
import `in`.rahulja.getlogs.model.LogType
import org.json.JSONObject

class LogLocationWorker @JvmOverloads constructor(
    appContext: Context,
    params: WorkerParameters,
    logRepo: LogRepository? = null,
    locationClient: FusedLocationProviderClient? = null
) : CoroutineWorker(appContext, params) {

    private val logRepository: LogRepository = logRepo ?: LogRepository.getInstance(appContext)
    private val fusedLocationClient: FusedLocationProviderClient =
        locationClient ?: LocationServices.getFusedLocationProviderClient(appContext)

    @Suppress("TooGenericExceptionCaught")
    override suspend fun doWork(): Result {
        val hasFine = ActivityCompat.checkSelfPermission(
            applicationContext,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED

        val hasCoarse = ActivityCompat.checkSelfPermission(
            applicationContext,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED

        if (!hasFine && !hasCoarse) {
            logRepository.recordLog(
                action = ACTION_LAST_LOCATION,
                dataPayload = JSONObject().apply {
                    put("error", "PERMISSION_DENIED")
                }.toString(),
                logType = LogType.LOCATION
            )
            return Result.failure()
        }

        return try {
            val location: Location? = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
                com.google.android.gms.tasks.Tasks.await(fusedLocationClient.lastLocation)
            }
            val payload = JSONObject().apply {
                if (location != null) {
                    put("latitude", location.latitude.toString())
                    put("longitude", location.longitude.toString())
                } else {
                    put("latitude", "")
                    put("longitude", "")
                }
            }.toString()

            logRepository.recordLog(
                action = ACTION_LAST_LOCATION,
                dataPayload = payload,
                logType = LogType.LOCATION
            )
            Result.success()
        } catch (e: Exception) {
            logRepository.recordLog(
                action = ACTION_LAST_LOCATION,
                dataPayload = JSONObject().apply {
                    put("error", e.message ?: "ERROR_GETTING")
                }.toString(),
                logType = LogType.LOCATION
            )
            Result.failure()
        }
    }

    companion object {
        const val ACTION_LAST_LOCATION = "in.rahulja.getlogs.LAST_LOCATION"
        const val TAG = "LogLocationWorker"
    }
}
